### PR TITLE
Fix generation of nested types that are not classes

### LINKED ISSUE #210: All source generators fail with nested types that are not classes
### Describe the bug

The source generator only assumes nested types will all be classes.
It should instead track the type of each nested type, and correclty generate that.

### Regression

No, always been broken

### Steps to reproduce

```text
Paste the following snippet:


public partial class MyClass
{
    public partial struct NestedStructType
    {
        public partial interface NestedInterfaceType
        {
            [ObservableRecipient]
            public partial class MyViewModel : ObservableValidator
            {
                [ObservableProperty]
                [Required]
                private string? name;

                [ICommand]
                private void MyCommand()
                {
                }
            }
        }
    }
}


This results in the following code:

```csharp
partial class MyClass
{
    partial class NestedStructType
    {
        partial class NestedInterfaceType
        {
            partial class MyViewModel
            {
                // Code here...
            }
        }
    }
}
```
```


### Expected behavior

The generated code would use the appropriate type for each nested type.

### Screenshots

_No response_

### IDE and version

VS 2022

### IDE version

17.1

### Nuget packages

- [ ] CommunityToolkit.Common
- [ ] CommunityToolkit.Diagnostics
- [ ] CommunityToolkit.HighPerformance
- [X] CommunityToolkit.Mvvm (aka MVVM Toolkit)

### Nuget package version(s)

main

### Additional context

_No response_

### Help us help you

Yes, I'd like to be assigned to work on this item

### PR DESCRIPTION
**Pull request by Sergio0694**
**Closes #210**

## PR Checklist

- [X] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [X] Based off latest main branch of toolkit
- [X] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [X] Tested code with current [supported SDKs](../#supported)
- [X] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Header has been added to all new source files (run _build/UpdateHeaders.bat_)
- [X] Contains **NO** breaking changes
- [X] Every new API (including internal ones) has full XML docs
- [X] Code follows all style conventions

### CODE REVIEW

File: tests/CommunityToolkit.Mvvm.UnitTests/Test_SourceGenerators.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                    private void Test()
+                    {
+                    }
+                }
+            }
+        }
+    }
+
+    [TestMethod]
+    public void Test_SourceGenerators_NestedTypesThatAreNotJustClassesAnddWithGenerics()
```
Comment (Arlodotexe) [inline tests/CommunityToolkit.Mvvm.UnitTests/Test_SourceGenerators.cs:52]: Minor typo

Comment (Arlodotexe) [inline tests/CommunityToolkit.Mvvm.UnitTests/Test_SourceGenerators.cs:52]:
```suggestion
    public void Test_SourceGenerators_NestedTypesThatAreNotJustClassesAndWithGenerics()
```

Comment (Arlodotexe) [inline tests/CommunityToolkit.Mvvm.UnitTests/Test_SourceGenerators.cs:52]: Minor typo

```suggestion
    public void Test_SourceGenerators_NestedTypesThatAreNotJustClassesAndWithGenerics()
```

**Review by Arlodotexe — APPROVED** [review]

File: CommunityToolkit.Mvvm.SourceGenerators/Models/HierarchyInfo.Syntax.cs [hunk 2]
```diff
@@ -30,28 +30,26 @@ partial record HierarchyInfo
         // Create the partial type declaration with the given member declarations.
         // This code produces a class declaration as follows:
         //
-        // partial class <TYPE_NAME>
+        // partial <TYPE_KIND> TYPE_NAME>
```
Comment (Nirmal4G) [inline CommunityToolkit.Mvvm.SourceGenerators/Models/HierarchyInfo.Syntax.cs:33]:
```suggestion
        // partial <TYPE_KIND> <TYPE_NAME>
```

**Pull request merged**